Make broadcast_object_list accept a device parameter.

bowangbj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61305**

Part I (this PR): Add dist_device argument to broadcast_object_list API
Part II: andwgu@ will deprecate _broadcast_object with the newly introduced API
	 Also include the changes to _object_to_tensor()/_tensor_to_object() with PR 60573

Context: https://github.com/pytorch/pytorch/issues/60062

Summary:

Test Plan:

Run the following on DevGpus with two cuda devices

$python setup.py develop    --- run this build on DevGPU
$BACKEND='nccl' WORLD_SIZE=2 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v
$BACKEND='gloo' WORLD_SIZE=2 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29566538](https://our.internmc.facebook.com/intern/diff/D29566538)

bowangbj committed:
Make broadcast_object_list accept a device parameter
Followed instruction at T93799916

Summary:

Test Plan:

Build with distributed on: USE_DISTRIBUTE=1 python setup.py develop
Test on CPU devvm:

$ with-proxy python test/distributed/optim/test_zero_redundancy_optimizer.py

Reviewers:

Subscribers:

Tasks:
T93799916

Tags:

[ghstack-poisoned]

mrshenli commented on test/distributed/optim/test_zero_redundancy_optimizer.py:
```diff
@@ -472,15 +472,17 @@ def closure():
-        optimizer_state_dict = _broadcast_object(
-            optimizer_state_dict,
+        object_list = [optimizer_state_dict]
+        optimizer_state_dict = distributed_c10d.broadcast_object_list(
```
curious, can this be called by running `dist.broadcast_object_list`?

bowangbj replied:
should be fine, was trying to make it consistent with other file, but makes more sense to use dist.distributed_c10d. directly since it only has one reference :)

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1700,7 +1700,7 @@ def gather_object(obj, object_gather_list=None, dst=0, group=None):
 
 
-def broadcast_object_list(object_list, src=0, group=None):
+def broadcast_object_list(object_list, src=0, group=None, dist_device=None):
```
shall we add test for this new argument?

bowangbj replied:
Added one test in test_broadcast_object_list

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1714,6 +1714,8 @@ def broadcast_object_list(object_list, src=0, group=None):
         src (int): Source rank from which to broadcast ``object_list``.
         group: (ProcessGroup, optional): The process group to work on. If None,
             the default process group will be used. Default is ``None``.
+        device (``torch.device``, optional): device to send from or receive
```
any reason this doc arg name does not math the API arg name?

bowangbj replied:
great catch

bowangbj committed:
draft: add dist_device with limited test on CPU on "Make broadcast_object_list accept a device parameter"


Followed instruction at T93799916

Summary:

Test Plan:

Build with distributed on: USE_DISTRIBUTE=1 python setup.py develop
Test on CPU devvm:

$ with-proxy python test/distributed/optim/test_zero_redundancy_optimizer.py

Reviewers:

Subscribers:

Tasks:
T93799916

Tags:

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29566538).

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1759,15 +1761,25 @@ def broadcast_object_list(object_list, src=0, group=None):
         object_sizes_tensor = torch.empty(len(object_list), dtype=torch.long)
 
+    # Current device selection.
+    # To preserve  backwards compatibility, dist_device is default to None in
```
duplicated space between `preserve` and  `backwards`

Also, please update `dist_device` accordingly if the arg name changed.

bowangbj replied:
done

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1759,15 +1761,25 @@ def broadcast_object_list(object_list, src=0, group=None):
+    # Current device selection.
+    # To preserve  backwards compatibility, dist_device is default to None in
+    # which case we run current logic of device selection, i.e. current_device
+    # is cuda if backend is NCCL otherwise CPU device. In the case it is not
```
cuda -> CUDA

bowangbj replied:
done

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1714,6 +1714,8 @@ def broadcast_object_list(object_list, src=0, group=None):
         group: (ProcessGroup, optional): The process group to work on. If None,
             the default process group will be used. Default is ``None``.
+        device (``torch.device``, optional): device to send from or receive
+            to (default: ``torch.device("cpu")``).
```
Shall we also explain the default behavior? The docstring says it defaults to CPU, but the code seems suggest that it will use current device if the backend is NCCL?

bowangbj replied:
Good point, updated. Was trying to minimize the change since I'm not familiar with the codebase yet :)

mrshenli reviewed (commented):
Thanks for putting this together!

mrshenli commented:
Hey @bowangbj, could you please update the PR summary to 1) remove internal-only info (e.g., task ID, proxy, devvm, etc.) 2) describe what's been changed and why 3) add "closes #60062"

mrshenli commented:
When you are ready, please get stamps from @rohan-varma and @andwgu :)

andwgu commented on torch/distributed/optim/zero_redundancy_optimizer.py:
```diff
@@ -298,29 +259,30 @@ def consolidate_state_dict(self, to: int = 0) -> None:
-                    local_state_dict = _broadcast_object(
-                        empty_messenger,
+                    object_list = [empty_messenger]
+                    distributed_c10d.broadcast_object_list(
```
Same as in the test file: I think `dist.broadcast_object_list()` works, so no need for the additional `import torch.distributed.distributed_c10d as distributed_c10d`.

bowangbj replied:
Got it, seems it's enabled by 

distributed/__init__.py 
  from .distributed_c10d import *  # noqa: F403 :)

andwgu commented on torch/distributed/optim/zero_redundancy_optimizer.py:
```diff
@@ -298,29 +259,30 @@ def consolidate_state_dict(self, to: int = 0) -> None:
                     # Receive the optimizer state from the source rank
-                    local_state_dict = _broadcast_object(
-                        empty_messenger,
+                    object_list = [empty_messenger]
```
Minor, but I do not think you need the `empty_messenger` anymore. You can set `object_list = [None]` and get rid of `empty_messenger`.

bowangbj replied:
Great point, revised as suggeted

andwgu commented on torch/distributed/optim/zero_redundancy_optimizer.py:
```diff
@@ -298,29 +259,30 @@ def consolidate_state_dict(self, to: int = 0) -> None:
+                        [self.optim.state_dict()],
                         src_rank=self.global_rank,
                         group=self.group,
                         device=self._default_device,
```
The parameter keyword above is `dist_device`, but `device` is used here.

bowangbj replied:
Great catch, renamed to dist_device

awgu reviewed (commented):
Thanks for taking the time to do this refactor. 

If you could check that [moving the received object tensor to `current_device`](https://github.com/pytorch/pytorch/blob/2a040475000e7046eab63757f561f354bf17badc/torch/distributed/distributed_c10d.py#L1798) before [calling `_tensor_to_object()`](https://github.com/pytorch/pytorch/blob/2a040475000e7046eab63757f561f354bf17badc/torch/distributed/distributed_c10d.py#L1809) suffices to avoid loading the tensors contained in the object to the source device, that would be great. This was the previous issue that was causing slowdowns. See [here](https://github.com/pytorch/pytorch/pull/60574). My solution was to change `_object_to_tensor()` and `_tensor_to_object()` to use `torch.save()` and `torch.load()` so I could specify a `map_location` argument in the load.

I worry that your current implementation correctly moves the received object tensor to the destination device, but then upon calling `_tensor_to_object()`, the tensors contained in the object (in the case of ZeRO, the tensors contained in the local state dict) are still loaded to the source device.

bowangbj committed:
resolve per Shen and Andrew comment on "Make broadcast_object_list accept a device parameter"


a) Add dist_device argument to broadcast_object_list API
b) deprecate _broadcast_object with the updated broadcast_object_list API

Context: https://github.com/pytorch/pytorch/issues/60062

Summary:

Test Plan:

- Build: $USE_DISTRIBUTED=1 USE_GLOO=1 python setup.py develop
- Test:
$python test/distributed/optim/test_zero_redundancy_optimizer.py --v
$BACKEND='gloo' WORLD_SIZE=4 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29566538](https://our.internmc.facebook.com/intern/diff/D29566538)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29566538).

bowangbj commented:
> Thanks for taking the time to do this refactor.
> 
> If you could check that [moving the received object tensor to `current_device`](https://github.com/pytorch/pytorch/blob/2a040475000e7046eab63757f561f354bf17badc/torch/distributed/distributed_c10d.py#L1798) before [calling `_tensor_to_object()`](https://github.com/pytorch/pytorch/blob/2a040475000e7046eab63757f561f354bf17badc/torch/distributed/distributed_c10d.py#L1809) suffices to avoid loading the tensors contained in the object to the source device, that would be great. This was the previous issue that was causing slowdowns. See [here](https://github.com/pytorch/pytorch/pull/60574). My solution was to change `_object_to_tensor()` and `_tensor_to_object()` to use `torch.save()` and `torch.load()` so I could specify a `map_location` argument in the load.
> 
> I worry that your current implementation correctly moves the received object tensor to the destination device, but then upon calling `_tensor_to_object()`, the tensors contained in the object (in the case of ZeRO, the tensors contained in the local state dict) are still loaded to the source device.



> Thanks for taking the time to do this refactor.
> 
> If you could check that [moving the received object tensor to `current_device`](https://github.com/pytorch/pytorch/blob/2a040475000e7046eab63757f561f354bf17badc/torch/distributed/distributed_c10d.py#L1798) before [calling `_tensor_to_object()`](https://github.com/pytorch/pytorch/blob/2a040475000e7046eab63757f561f354bf17badc/torch/distributed/distributed_c10d.py#L1809) suffices to avoid loading the tensors contained in the object to the source device, that would be great. This was the previous issue that was causing slowdowns. See [here](https://github.com/pytorch/pytorch/pull/60574). My solution was to change `_object_to_tensor()` and `_tensor_to_object()` to use `torch.save()` and `torch.load()` so I could specify a `map_location` argument in the load.
> 
> I worry that your current implementation correctly moves the received object tensor to the destination device, but then upon calling `_tensor_to_object()`, the tensors contained in the object (in the case of ZeRO, the tensors contained in the local state dict) are still loaded to the source device.

May I do that as a follow up since this is quite different from the issue https://github.com/pytorch/pytorch/issues/60062 I am trying to resolve, and I am quite new to the code base (bootcamp task) :)

bowangbj commented:
Thanks Shen and Andrew a lot for your super quick review. Resolve most of the cmts, PTAL

andwgu commented:
Thanks for making the changes!

> May I do that as a follow up since this is quite different from the issue #60062 I am trying to resolve, and I am quite new to the code base (bootcamp task) :)

I see. Since I am an intern, I am not familiar with the bootcamp process (e.g. how fast the deadlines are), but if you can make this change in a follow-up PR then that should be fine. Though, we should check with @mrshenli and @rohan-varma.

To reiterate, my suspicion is that this PR change does not address the issue that tensors contained inside broadcasted objects (e.g. `ZeroRedundancyOptimizer`'s local state dict) are loaded to the source device instead of the destination device due to the default behavior of `Unpickler.load()`. My concern is that merging this PR then will actually not preserve true backwards compatibility in the sense that the internal behavior of `ZeroRedundancyOptimizer` will change and result in a slowdown (i.e. `_broadcast_object()` and `broadcast_object_list()` with this added `dist_device` argument are not actually directly interchangeable). Practically, this just means that between now and the follow-up PR, `ZeroRedundancyOptimizer` will have a greater latency when calling `consolidate_state_dict()` (due to the data transfer from source device to destination device). 

If there is urgency to merge this PR, then that is fine. Otherwise, I feel that most of the work to make the change with `_object_to_tensor()` and `_tensor_to_object()` is already there in my [open PR](https://github.com/pytorch/pytorch/pull/60574). We just need to agree on what further tests need to be written to solidify the change. If you do not have the bandwidth to do the follow-up PR, I can do it as well. I do not anticipate that it should require too much additional work.

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1714,6 +1715,9 @@ def broadcast_object_list(object_list, src=0, group=None):
         src (int): Source rank from which to broadcast ``object_list``.
         group: (ProcessGroup, optional): The process group to work on. If None,
             the default process group will be used. Default is ``None``.
+        dist_device (``torch.device``, optional): device to send from or receive
```
Would be good to specify that if this is not `None`, then intermediate tensor representations of objects, or any tensors within the contained objects, will be moved to this device before broadcasting. 

Also mention that default for this arg is None.

bowangbj replied:
Good point, revised.

rohan-varma reviewed (commented):
I think @andwgu's suggestion is reasonable, and it's fine to do in a follow up PR, but can we revert the `zero` changes here and constrain this PR to only affecting `broadcast_object_list` API? Currently it also makes ZeRO adopt this API but that would ideally be part of a separate PR. In addition @andwgu 's related change https://github.com/pytorch/pytorch/pull/60573#pullrequestreview-701538406 does something similar but only touches the `map_location` and still uses current_device semantics (which this PR changes). Based on that we can land the distributed_c10d changes here, and then incorporate Andrew's PR on top?

andwgu commented:
> I think @andwgu's suggestion is reasonable, and it's fine to do in a follow up PR, but can we revert the `zero` changes here and constrain this PR to only affecting `broadcast_object_list` API? Currently it also makes ZeRO adopt this API but that would ideally be part of a separate PR. In addition @andwgu 's related change [#60573 (review)](https://github.com/pytorch/pytorch/pull/60573#pullrequestreview-701538406) does something similar but only touches the `map_location` and still uses current_device semantics (which this PR changes). Based on that we can land the distributed_c10d changes here, and then incorporate Andrew's PR on top?

Commenting to say that I am on board with @rohan-varma's plan.
- @bowangbj changes this PR to only affect `broadcast_object_list()`, introducing the new argument `dist_device` and changing the `current_device` semantics.
- In a follow-up PR, we can include the changes to `_object_to_tensor()`/`_tensor_to_object()` and the migrating of `_broadcast_object()` to `broadcast_object_list()` in `ZeroRedundancyOptimizer`. If there is no issue, I can do this PR.

bowangbj committed:
update per rohan and andrew cmt on "Make broadcast_object_list accept a device parameter."


Part I (this PR): Add dist_device argument to broadcast_object_list API
Part II: andwgu@ will deprecate _broadcast_object with the newly introduced API
	 Also include the changes to _object_to_tensor()/_tensor_to_object() with PR 60573

Context: https://github.com/pytorch/pytorch/issues/60062

Summary:

Test Plan:

$USE_DISTRIBUTED=1 USE_GLOO=1 python setup.py develop
$BACKEND='gloo' WORLD_SIZE=4 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29566538](https://our.internmc.facebook.com/intern/diff/D29566538)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29566538).

bowangbj committed:
small lint on "Make broadcast_object_list accept a device parameter."


Part I (this PR): Add dist_device argument to broadcast_object_list API
Part II: andwgu@ will deprecate _broadcast_object with the newly introduced API
	 Also include the changes to _object_to_tensor()/_tensor_to_object() with PR 60573

Context: https://github.com/pytorch/pytorch/issues/60062

Summary:

Test Plan:

$USE_DISTRIBUTED=1 USE_GLOO=1 python setup.py develop
$BACKEND='gloo' WORLD_SIZE=4 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29566538](https://our.internmc.facebook.com/intern/diff/D29566538)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29566538).

bowangbj committed:
small lint on "Make broadcast_object_list accept a device parameter."


Part I (this PR): Add dist_device argument to broadcast_object_list API
Part II: andwgu@ will deprecate _broadcast_object with the newly introduced API
	 Also include the changes to _object_to_tensor()/_tensor_to_object() with PR 60573

Context: https://github.com/pytorch/pytorch/issues/60062

Summary:

Test Plan:

$USE_DISTRIBUTED=1 USE_GLOO=1 python setup.py develop
$BACKEND='gloo' WORLD_SIZE=4 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29566538](https://our.internmc.facebook.com/intern/diff/D29566538)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29566538).

bowangbj commented:
> > I think @andwgu's suggestion is reasonable, and it's fine to do in a follow up PR, but can we revert the `zero` changes here and constrain this PR to only affecting `broadcast_object_list` API? Currently it also makes ZeRO adopt this API but that would ideally be part of a separate PR. In addition @andwgu 's related change [#60573 (review)](https://github.com/pytorch/pytorch/pull/60573#pullrequestreview-701538406) does something similar but only touches the `map_location` and still uses current_device semantics (which this PR changes). Based on that we can land the distributed_c10d changes here, and then incorporate Andrew's PR on top?
> 
> Commenting to say that I am on board with @rohan-varma's plan.
> 
> * @bowangbj changes this PR to only affect `broadcast_object_list()`, introducing the new argument `dist_device` and changing the `current_device` semantics.
> * In a follow-up PR, we can include the changes to `_object_to_tensor()`/`_tensor_to_object()` and the migrating of `_broadcast_object()` to `broadcast_object_list()` in `ZeroRedundancyOptimizer`. If there is no issue, I can do this PR.

Thanks Rohan and Andrew for the great comment. 
Also thanks Andrew for the great help to take the follow up. 
Updated the CL to cover API change to broadcast_object_list only (with backward compatibility). 
PTAL

andwgu commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1714,6 +1714,10 @@ def broadcast_object_list(object_list, src=0, group=None):
         src (int): Source rank from which to broadcast ``object_list``.
         group: (ProcessGroup, optional): The process group to work on. If None,
             the default process group will be used. Default is ``None``.
+        dist_device (``torch.device``, optional): If nont None, then intermediate
```
nit: `nont` -> `not` and maybe change `None` to be \`\` `None` ``.

bowangbj replied:
great catch.

andwgu commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1759,15 +1765,25 @@ def broadcast_object_list(object_list, src=0, group=None):
         object_sizes_tensor = torch.empty(len(object_list), dtype=torch.long)
 
+    # Current device selection.
+    # To preserve backwards compatibility, dist_device is default to None in
```
nit: might also want to add \`\` \`\` around `dist_device`, `None`, and `current_device` in this paragraph.

bowangbj replied:
SG

awgu reviewed (approved):
I added two minor style-related comments. Otherwise, LGTM! I will mark as approved, but I am not sure if I am supposed to have approval power. Perhaps wait until one of the others takes another look.

bowangbj committed:
resolve andrew comment on "Make broadcast_object_list accept a device parameter."


Part I (this PR): Add dist_device argument to broadcast_object_list API
Part II: andwgu@ will deprecate _broadcast_object with the newly introduced API
	 Also include the changes to _object_to_tensor()/_tensor_to_object() with PR 60573

Context: https://github.com/pytorch/pytorch/issues/60062

Summary:

Test Plan:

$USE_DISTRIBUTED=1 USE_GLOO=1 python setup.py develop
$BACKEND='gloo' WORLD_SIZE=4 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29566538](https://our.internmc.facebook.com/intern/diff/D29566538)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29566538).

bowangbj commented:
> I added two minor style-related comments. Otherwise, LGTM! I will mark as approved, but I am not sure if I am supposed to have approval power. Perhaps wait until one of the others takes another look.

Thanks Andrew for your super fast and great review, resolved all the style issues. Seems now I can land my 1st github code change :) 

@rohan-varma @mrshenli  wanted to take another quick look before I merge? 

Thanks

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1714,6 +1714,10 @@ def broadcast_object_list(object_list, src=0, group=None):
         src (int): Source rank from which to broadcast ``object_list``.
         group: (ProcessGroup, optional): The process group to work on. If None,
             the default process group will be used. Default is ``None``.
+        dist_device (``torch.device``, optional): If not None, then intermediate
```
Curious, any reason for choosing the name `dist_device` instead of `device` or `map_location` or other names? Is the `dist_`  prefix here stands for distributed? If so, why this is an "distributed device"?

rohan-varma replied:
+1, maybe just call it `device`?

bowangbj replied:
mainly b/c dist_device was mentioned here https://github.com/pytorch/pytorch/issues/60062 :) 
changed it to device.

mrshenli commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5778,7 +5778,14 @@ def test_broadcast_object_list(self):
             )
 
-            # Single object test
+            # Single object test with dist_device_specified
```
dist_device_specified -> dist_device specified

bowangbj replied:
great catch

mrshenli reviewed (approved):
LGTM too! Thanks @bowangbj! Left two minor comments.

mrshenli reviewed (changes requested):
Oh test failures are real:

```
Jul 08 18:54:08 ======================================================================
Jul 08 18:54:08 ERROR [3.894s]: test_broadcast_object_list (__main__.TestDistBackendWithSpawn)
Jul 08 18:54:08 ----------------------------------------------------------------------
Jul 08 18:54:08 Traceback (most recent call last):
Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_distributed.py", line 398, in wrapper
Jul 08 18:54:08     self._join_processes(fn)
Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_distributed.py", line 610, in _join_processes
Jul 08 18:54:08     self._check_return_codes(elapsed_time)
Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_distributed.py", line 653, in _check_return_codes
Jul 08 18:54:08     raise RuntimeError(error)
Jul 08 18:54:08 RuntimeError: Process 1 exited with error code 10 and exception:
Jul 08 18:54:08 Traceback (most recent call last):
Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_distributed.py", line 529, in run_test
Jul 08 18:54:08     getattr(self, test_name)()
Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_distributed.py", line 400, in wrapper
Jul 08 18:54:08     fn()
Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_distributed.py", line 82, in wrapper
Jul 08 18:54:08     return func(*args, **kwargs)
Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/distributed/distributed_test.py", line 5785, in test_broadcast_object_list
Jul 08 18:54:08     dist.broadcast_object_list(single_obj_list, src=0, group=None, dist_device=torch.device("cpu"))
Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/distributed/distributed_c10d.py", line 1790, in broadcast_object_list
Jul 08 18:54:08     broadcast(object_sizes_tensor, src=src, group=group)
Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/distributed/distributed_c10d.py", line 1136, in broadcast
Jul 08 18:54:08     work = default_pg.broadcast([tensor], opts)
Jul 08 18:54:08 RuntimeError: Tensors must be CUDA and dense
```

andwgu commented:
> Oh test failures are real:
> 
> ```
> Jul 08 18:54:08 ======================================================================
> Jul 08 18:54:08 ERROR [3.894s]: test_broadcast_object_list (__main__.TestDistBackendWithSpawn)
> Jul 08 18:54:08 ----------------------------------------------------------------------
> Jul 08 18:54:08 Traceback (most recent call last):
> Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_distributed.py", line 398, in wrapper
> Jul 08 18:54:08     self._join_processes(fn)
> Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_distributed.py", line 610, in _join_processes
> Jul 08 18:54:08     self._check_return_codes(elapsed_time)
> Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_distributed.py", line 653, in _check_return_codes
> Jul 08 18:54:08     raise RuntimeError(error)
> Jul 08 18:54:08 RuntimeError: Process 1 exited with error code 10 and exception:
> Jul 08 18:54:08 Traceback (most recent call last):
> Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_distributed.py", line 529, in run_test
> Jul 08 18:54:08     getattr(self, test_name)()
> Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_distributed.py", line 400, in wrapper
> Jul 08 18:54:08     fn()
> Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_distributed.py", line 82, in wrapper
> Jul 08 18:54:08     return func(*args, **kwargs)
> Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/distributed/distributed_test.py", line 5785, in test_broadcast_object_list
> Jul 08 18:54:08     dist.broadcast_object_list(single_obj_list, src=0, group=None, dist_device=torch.device("cpu"))
> Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/distributed/distributed_c10d.py", line 1790, in broadcast_object_list
> Jul 08 18:54:08     broadcast(object_sizes_tensor, src=src, group=group)
> Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/distributed/distributed_c10d.py", line 1136, in broadcast
> Jul 08 18:54:08     work = default_pg.broadcast([tensor], opts)
> Jul 08 18:54:08 RuntimeError: Tensors must be CUDA and dense
> ```

The issue I think is that NCCL backend requires tensors to be on GPU before broadcasting, so in your test case when you specify `dist_device=torch.device("cpu")`, it will lead to the error above. There are some ways to approach this. One way I think is to pass in `dist_device=self.device` and add a `.to(torch.device("cpu"))` before comparing `assertEquals()`.

andwgu commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1714,6 +1714,10 @@ def broadcast_object_list(object_list, src=0, group=None):
         group: (ProcessGroup, optional): The process group to work on. If None,
             the default process group will be used. Default is ``None``.
+        dist_device (``torch.device``, optional): If not None, then intermediate
+            tensor representations of objects, or any tensors within the contained
```
> "or any tensors within the contained objects, will be moved to this device before broadcasting"

Sorry for missing this before. Upon closer examination, I do not believe this is true. The size and object tensors are indeed moved to `dist_device` before broadcasting, but those two tensors are the result of serializing the inputted objects. I do not think that given the current implementation of `_object_to_tensor()` and `_tensor_to_object()`, the tensors contained in the object pre-serialization will be moved to `dist_device`.

The `dist_device` argument is just the device used for the broadcast. Hence, I think precisely, this PR simply adds an argument that allows a process to broadcast from a specified device. 
Before:
- If the backend is NCCL, then the broadcast uses `torch.cuda.current_device()`.
- If the backend is not NCCL, then it uses `torch.device("cpu")`.

Now:
- If the backend is NCCL, a process can use something other than `torch.cuda.current_device()` (but that something must still be a GPU).
- If the backend is not NCCL, a process can use something other than CPU. 

Given this realization, I wonder how useful this change is. @rohan-varma @mrshenli 
- If the backend is NCCL, why not use `torch.cuda.current_device()`? Even if there exists a second GPU available, why would that be any better to use? 
- If the backend is not NCCL, then would be using GPU for the broadcast be desirable? I think for both GLOO and MPI backends, using CPU is preferred (correct me if I am wrong).

rohan-varma replied:
I think it's useful to have the flexibility when backend is NCCL to control which cuda device tensors are placed on. It makes it more similar to `dist.broadcast(tensor)` APIs where the tensor can be on a device which is not necessarily the current device. This use case might be rare but I don't see a disadvantage to enabling it.

Regarding Gloo backend, Gloo is also frequently used with GPU, although for this particular API, I suspect perf would be better if the user simply used the CPU. But I'm not sure if that's something we should enforce and it would make the API more prescriptive. Given that I think it's okay to respect the passed in device being a GPU for gloo backend, if CUDA is enabled.

bowangbj replied:
Given rohan's comment, WDYT Andrew? Sorry I was not quite familiar with the code, will defer to both of you :)

andwgu replied:
I think what @rohan-varma said makes sense. Let us move forward with adding this `device` argument, which specifies which device the size and object tensors are moved to before broadcasting. 

In my follow-up PR (which should be ready by tomorrow), I plan to introduce a `map_location` argument, which specifies which device the tensors in the broadcasted object will be loaded to upon being received.

bowangbj replied:
SG

andwgu replied:
Minor comment:

> or any tensors within the contained objects, will be moved to this device before broadcasting.

This wording in the comment may still be misleading since it glosses over the serialization step. At least to me, the comment suggests that the tensors contained in the objects are moved to `device` and are themselves broadcasted; however, actually the object is serialized and converted to a tensor. That tensor is moved to `device` and broadcasted.

bowangbj replied:
updated per your last sentence, how does it sound?

awgu reviewed (changes requested):
I apologize @bowangbj for all the confusion. I took another look and had some comments (not really directed to you).

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1759,15 +1765,25 @@ def broadcast_object_list(object_list, src=0, group=None):
+    # To preserve backwards compatibility, ``dist_device`` is default to ``None``
+    # in which case we run current logic of device selection, i.e.
+    # ``current_device`` is CUDA if backend is NCCL otherwise CPU device. In the
+    # case it is not ``None``e we move all intermediate tensors to this given
```
nit: typo (``None``e)

bowangbj replied:
Great catch

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5778,7 +5778,14 @@ def test_broadcast_object_list(self):
+            if self.rank != src_rank:
+                self.assertNotEqual(single_obj_list[0], COLLECTIVES_OBJECT_TEST_LIST[0])
+            dist.broadcast_object_list(single_obj_list, src=0, group=None, dist_device=torch.device("cpu"))
+            self.assertEqual(single_obj_list[0], COLLECTIVES_OBJECT_TEST_LIST[0])
```
can we also have a test where we specify a non-current CUDA device for both nccl and gloo backends?

bowangbj replied:
updated current test to use next rank

andwgu commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1759,15 +1765,25 @@ def broadcast_object_list(object_list, src=0, group=None):
-    current_device = torch.device("cpu")
+    current_device = None
+    if dist_device is not None:
+        current_device = dist_device
```
Given the discussion [here](https://github.com/pytorch/pytorch/pull/61305#discussion_r666599220), since NCCL backend requires the device to be a CUDA device, I think it would be better to check here that `dist_device` is a CUDA device (e.g. `dist_device.type == "cuda"`) if `is_nccl_backend` and explicitly raise an error (e.g. `ValueError`) if not to avoid erroring silently.

bowangbj replied:
Got it, added a check

bowangbj committed:
resolve cmt from andrew/shen/rohan on "Make broadcast_object_list accept a device parameter."


Part I (this PR): Add dist_device argument to broadcast_object_list API
Part II: andwgu@ will deprecate _broadcast_object with the newly introduced API
	 Also include the changes to _object_to_tensor()/_tensor_to_object() with PR 60573

Context: https://github.com/pytorch/pytorch/issues/60062

Summary:

Test Plan:

$USE_DISTRIBUTED=1 USE_GLOO=1 python setup.py develop
$BACKEND='gloo' WORLD_SIZE=4 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29566538](https://our.internmc.facebook.com/intern/diff/D29566538)

[ghstack-poisoned]

bowangbj committed:
update summary on "Make broadcast_object_list accept a device parameter."


Part I (this PR): Add dist_device argument to broadcast_object_list API
Part II: andwgu@ will deprecate _broadcast_object with the newly introduced API
	 Also include the changes to _object_to_tensor()/_tensor_to_object() with PR 60573

Context: https://github.com/pytorch/pytorch/issues/60062

Summary:

Test Plan:

$python setup.py develop    --- run this build on DevGPU
$BACKEND='nccl' WORLD_SIZE=4 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29566538](https://our.internmc.facebook.com/intern/diff/D29566538)

[ghstack-poisoned]

bowangbj commented:
> Oh test failures are real:
> 
> ```
> Jul 08 18:54:08 ======================================================================
> Jul 08 18:54:08 ERROR [3.894s]: test_broadcast_object_list (__main__.TestDistBackendWithSpawn)
> Jul 08 18:54:08 ----------------------------------------------------------------------
> Jul 08 18:54:08 Traceback (most recent call last):
> Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_distributed.py", line 398, in wrapper
> Jul 08 18:54:08     self._join_processes(fn)
> Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_distributed.py", line 610, in _join_processes
> Jul 08 18:54:08     self._check_return_codes(elapsed_time)
> Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_distributed.py", line 653, in _check_return_codes
> Jul 08 18:54:08     raise RuntimeError(error)
> Jul 08 18:54:08 RuntimeError: Process 1 exited with error code 10 and exception:
> Jul 08 18:54:08 Traceback (most recent call last):
> Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_distributed.py", line 529, in run_test
> Jul 08 18:54:08     getattr(self, test_name)()
> Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_distributed.py", line 400, in wrapper
> Jul 08 18:54:08     fn()
> Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/common_distributed.py", line 82, in wrapper
> Jul 08 18:54:08     return func(*args, **kwargs)
> Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/testing/_internal/distributed/distributed_test.py", line 5785, in test_broadcast_object_list
> Jul 08 18:54:08     dist.broadcast_object_list(single_obj_list, src=0, group=None, dist_device=torch.device("cpu"))
> Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/distributed/distributed_c10d.py", line 1790, in broadcast_object_list
> Jul 08 18:54:08     broadcast(object_sizes_tensor, src=src, group=group)
> Jul 08 18:54:08   File "/opt/conda/lib/python3.9/site-packages/torch/distributed/distributed_c10d.py", line 1136, in broadcast
> Jul 08 18:54:08     work = default_pg.broadcast([tensor], opts)
> Jul 08 18:54:08 RuntimeError: Tensors must be CUDA and dense
> ```

Oops, I guess it's failed b/c I was using devvm. Switched to devgpus and  set backend to "nccl" to test it.

bowangbj commented:
Thanks Andrew, Rohan and Shen for all the great and quick review, finally I got a devgpu device and verified it on it, updated all the cmts, PTAL

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29566538).

andwgu commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1759,15 +1765,27 @@ def broadcast_object_list(object_list, src=0, group=None):
+    # in which case we run current logic of device selection, i.e.
+    # ``current_device`` is CUDA if backend is NCCL otherwise CPU device. In the
+    # case it is not ``None`` we move all intermediate tensors to this given
+    # device. See https://github.com/pytorch/pytorch/issues/60062
```
> we move all intermediate tensors to this given device

nit: Just as in this [comment](https://github.com/pytorch/pytorch/pull/61305/files#r667955832), I personally feel that `intermediate tensors` may be misleading. It may be better to say `we move the size and object tensors to be broadcasted to this device`.

andwgu replied:
nit: Also, should we leave the Github issue link here, especially if the issue will be closed post-merge? I feel that ideally the documentation here is sufficiently self-explanatory, which I think it already is.

bowangbj replied:
SG, done

andwgu commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5778,7 +5778,15 @@ def test_broadcast_object_list(self):
 
-            # Single object test
+            # Single object test with device specified
+            if backend == "nccl":
```
@rohan-varma [requested](https://github.com/pytorch/pytorch/pull/61305#discussion_r666599812) a test for non-current CUDA device for both `nccl` and `gloo` backends, but this test only covers `nccl` backend.

I think the `if backend == "nccl":` can probably be removed, and the test should still pass since `gloo` works with GPU as well as mentioned by @rohan-varma [in this comment](https://github.com/pytorch/pytorch/pull/61305#discussion_r666599220).

bowangbj replied:
Good point, reverted the if statement. 

dumb Q, not quite familiar with this test. The way I run it is 

for gloo
- BACKEND='gloo' WORLD_SIZE=4 with-proxy python test/distributed/test_distributed_fork.py TestDistBackendWithFork.test_broadcast_object_list --v

for nccl

BACKEND='nccl' WORLD_SIZE=4 with-proxy python test/distributed/test_distributed_fork.py TestDistBackendWithFork.test_broadcast_object_list --v


is that enough?

andwgu replied:
I am also not fully familiar with the distributed testing, but since `backend` is read from the environment variable, I am fairly sure that the way you tested is correct.

I would just wait to see how the CI tests go since I am not sure if the test is ever run with Gloo backend with no GPUs available. In that case, I think that this [line](https://github.com/pytorch/pytorch/blob/1eb70690dcd975bc6ae006bda2406126eb4ac3a8/torch/testing/_internal/distributed/distributed_test.py#L5786) in your test will error due to referencing a non-existing CUDA device.

If I were you, I would duplicate the test you added so that there are two versions: (1) using `torch.device("cpu")` as the specified `device` and (2) using `next_rank` as the specified `device`. In particular, only run (2) if `torch.cuda.device_count() == dist.get_world_size()`, so you know for sure that the line mentioned above will not error.

bowangbj replied:
Great idea, I assume the tests executed twice in two different env variable backend=gloo vs backend=nccl. 
Just setting the device conditionally should be enough
device = torch.device('cpu') if backend == 'gloo' else torch.device(next_rank)

andwgu replied:
IIUC, this does not test Gloo with GPU then right? (only NCCL with GPU and Gloo with CPU)

Quoting from @rohan-varma [here](https://github.com/pytorch/pytorch/pull/61305#discussion_r666599220):

> Regarding Gloo backend, Gloo is also frequently used with GPU

I was thinking that if a main part of the motivation for adding this `device` argument is so that users can specify GPU for Gloo backend, then we should test it.

bowangbj replied:
Got it, duplicates the test a few times to make sure we can test gloo -- cpu and gpu (next_rank) as well as nccl - gpu(next_rank). How does current version sound like?

rohan-varma replied:
+1, It would be great if GPU device could be supported in Gloo and a UT for it.

bowangbj replied:
done, added three tests here

bowangbj committed:
resolve andrew cmt on "Make broadcast_object_list accept a device parameter."


Part I (this PR): Add dist_device argument to broadcast_object_list API
Part II: andwgu@ will deprecate _broadcast_object with the newly introduced API
	 Also include the changes to _object_to_tensor()/_tensor_to_object() with PR 60573

Context: https://github.com/pytorch/pytorch/issues/60062

Summary:

Test Plan:

$python setup.py develop    --- run this build on DevGPU
$BACKEND='nccl' WORLD_SIZE=4 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29566538](https://our.internmc.facebook.com/intern/diff/D29566538)

[ghstack-poisoned]

bowangbj commented:
Thanks Andrew for your super detailed review, resolved most of the cmts, pTAL

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29566538).

andwgu commented:
> Thanks Andrew for your super detailed review, resolved most of the cmts, pTAL

No problem. This is the last concern from me: https://github.com/pytorch/pytorch/pull/61305#discussion_r668146137. After we resolve this, the PR should be good to go.

bowangbj committed:
ensure test compatible with gloo and nccl on "Make broadcast_object_list accept a device parameter."


Part I (this PR): Add dist_device argument to broadcast_object_list API
Part II: andwgu@ will deprecate _broadcast_object with the newly introduced API
	 Also include the changes to _object_to_tensor()/_tensor_to_object() with PR 60573

Context: https://github.com/pytorch/pytorch/issues/60062

Summary:

Test Plan:

$python setup.py develop    --- run this build on DevGPU
$BACKEND='nccl' WORLD_SIZE=4 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29566538](https://our.internmc.facebook.com/intern/diff/D29566538)

[ghstack-poisoned]

bowangbj committed:
update pr summary on "Make broadcast_object_list accept a device parameter."


Part I (this PR): Add dist_device argument to broadcast_object_list API
Part II: andwgu@ will deprecate _broadcast_object with the newly introduced API
	 Also include the changes to _object_to_tensor()/_tensor_to_object() with PR 60573

Context: https://github.com/pytorch/pytorch/issues/60062

Summary:

Test Plan:

$python setup.py develop    --- run this build on DevGPU
$BACKEND='nccl' WORLD_SIZE=2 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v
$BACKEND='gloo' WORLD_SIZE=2 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29566538](https://our.internmc.facebook.com/intern/diff/D29566538)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29566538).

bowangbj commented:
> > Thanks Andrew for your super detailed review, resolved most of the cmts, pTAL
> 
> No problem. This is the last concern from me: [#61305 (comment)](https://github.com/pytorch/pytorch/pull/61305#discussion_r668146137). After we resolve this, the PR should be good to go.

Thanks Andrew, updated the test as suggested, PTAL

bowangbj committed:
duplicate tests for gloo and nccl on "Make broadcast_object_list accept a device parameter."


Part I (this PR): Add dist_device argument to broadcast_object_list API
Part II: andwgu@ will deprecate _broadcast_object with the newly introduced API
	 Also include the changes to _object_to_tensor()/_tensor_to_object() with PR 60573

Context: https://github.com/pytorch/pytorch/issues/60062

Summary:

Test Plan:

$python setup.py develop    --- run this build on DevGPU
$BACKEND='nccl' WORLD_SIZE=2 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v
$BACKEND='gloo' WORLD_SIZE=2 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29566538](https://our.internmc.facebook.com/intern/diff/D29566538)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29566538).

andwgu commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5778,13 +5778,29 @@ def test_broadcast_object_list(self):
+                dist.broadcast_object_list(single_obj_list, src=0, group=None, device=torch.device('cpu'))
+                self.assertEqual(single_obj_list[0], COLLECTIVES_OBJECT_TEST_LIST[0])
+
+            # # Single object test with device specified. Backend="gloo", device=current_device+1
```
nit: Double `#` here.

bowangbj replied:
good catch

andwgu commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5778,13 +5778,29 @@ def test_broadcast_object_list(self):
+                single_obj_list = [objects[0]]
+                if self.rank != src_rank:
+                    self.assertNotEqual(single_obj_list[0], COLLECTIVES_OBJECT_TEST_LIST[0])
+                dist.broadcast_object_list(single_obj_list, src=0, group=None, device=torch.device(next_rank))
```
I mentioned previously:
> only run (2) if torch.cuda.device_count() == dist.get_world_size() to ensure that the line mentioned above will not error. 

I prefer to have some kind of check here to be defensive against varying test hardware configurations (e.g. if CI changes its setup, someone runs these tests locally with their own setup, etc.). This line with `device=torch.device(next_rank)` will error if there are no GPUs available. Currently, I think the testing ensures the invariant that if `backend="nccl"`, then there will be GPUs available, but I do not believe that that same guarantee is made for `backend="gloo"`. That is why I suggested to check that `torch.cuda.device_count() == dist.get_world_size()`. This guarantees that the scheme of setting the device to be `rank + 1` will not error. If you want to check the GPU availability in another way, feel free to do so.

mrshenli replied:
+1 to @andwgu's comment. Another option is extracting this test to its own unit test function, and add decorators like `@skip_if_lt_x()` accordingly.

https://github.com/pytorch/pytorch/blob/8a2c7d902fb159bb57a0cf46f3a5ba74095cc34a/torch/testing/_internal/distributed/distributed_test.py#L875

rohan-varma replied:
@andwgu The test already has the following decorator - 
```
@require_n_gpus_for_nccl_backend(
            int(os.environ["WORLD_SIZE"]), os.environ["BACKEND"]
        )
```

shouldn't that be enough to ensure there are no errors here due to no GPUs available?

mrshenli replied:
Looks like this decorator only checks GPUs when the backend is NCCL, but not gloo?

https://github.com/pytorch/pytorch/blob/14f63763c16e77ac4793c938003e791dd71b217d/torch/testing/_internal/common_distributed.py#L79-L89

bowangbj replied:
Very great point! 
leveraged Andrew's suggestion here
if backend != "nccl" and torch.cuda.device_count() == int(self.world_size)

mrshenli commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5761,10 +5761,10 @@ def forward(self, _):
         def test_broadcast_object_list(self):
             # Only set device for NCCL backend since it must use GPUs.
+            # Case where rank != GPU device.
+            next_rank = (self.rank + 1) % int(self.world_size)
```
curious, is `int()` here necessary?

bowangbj replied:
self.world_size is typed str.  thus int is needed.

mrshenli commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5778,7 +5778,31 @@ def test_broadcast_object_list(self):
 
-            # Single object test
+            # Single object test with device specified. Backend="gloo", device=cpu
+            if backend == "gloo":
```
Curious, will `backend != "nccl"` work here? If so, that might be better, as it will also give us coverage for MPI backend.

bowangbj replied:
in the header 
@require_backend({"nccl", "gloo"}) is specified
does this mean only nccl and gloo backends are valid for this test?

anyway, changing it to backend != nccl should work

mrshenli commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5778,7 +5778,31 @@ def test_broadcast_object_list(self):
+                single_obj_list = [objects[0]]
+                if self.rank != src_rank:
+                    self.assertNotEqual(single_obj_list[0], COLLECTIVES_OBJECT_TEST_LIST[0])
+                dist.broadcast_object_list(single_obj_list, src=0, group=None, device=torch.device(next_rank))
```
ditto. Even for NCCL, it's possible that the number of available devices is less then world size.

bowangbj replied:
added check
torch.cuda.device_count() == int(self.world_size) 

as before

mrshenli reviewed (commented):
Looks great! Thanks! We just need a few minor touches before landing.

mrshenli reviewed (commented):
Hey @bowangbj, in your next update. Could you please rebase this PR to avoid merge conflicts in the CI? Thanks!

bowangbj committed:
update test on "Make broadcast_object_list accept a device parameter."


Part I (this PR): Add dist_device argument to broadcast_object_list API
Part II: andwgu@ will deprecate _broadcast_object with the newly introduced API
	 Also include the changes to _object_to_tensor()/_tensor_to_object() with PR 60573

Context: https://github.com/pytorch/pytorch/issues/60062

Summary:

Test Plan:

$python setup.py develop    --- run this build on DevGPU
$BACKEND='nccl' WORLD_SIZE=2 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v
$BACKEND='gloo' WORLD_SIZE=2 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29566538](https://our.internmc.facebook.com/intern/diff/D29566538)

[ghstack-poisoned]

bowangbj committed:
update device setup on "Make broadcast_object_list accept a device parameter."


Part I (this PR): Add dist_device argument to broadcast_object_list API
Part II: andwgu@ will deprecate _broadcast_object with the newly introduced API
	 Also include the changes to _object_to_tensor()/_tensor_to_object() with PR 60573

Context: https://github.com/pytorch/pytorch/issues/60062

Summary:

Test Plan:

Run the following on DevGpus with two cuda devices

$python setup.py develop    --- run this build on DevGPU
$BACKEND='nccl' WORLD_SIZE=2 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v
$BACKEND='gloo' WORLD_SIZE=2 with-proxy  python test/distributed/test_distributed_fork.py  TestDistBackendWithFork.test_broadcast_object_list --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29566538](https://our.internmc.facebook.com/intern/diff/D29566538)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29566538).

bowangbj commented:
> Looks great! Thanks! We just need a few minor touches before landing.

Rebased to HEAD.

Thanks Shen, Andrew and Rohan for the great and detailed review. Resolved all the cmts. Verified the test cases on DevGPU. PTAL.

mrshenli reviewed (approved):
LGTM! Thank you @bowangbj

bowangbj commented:
> LGTM! Thank you @bowangbj

Thanks Shen and all for the great review. Merged the pR, if any further comment, I will address it as follow up. Andrew, now, t's ready to handle over to you for part II :)
